### PR TITLE
Add flag to check for uninitialized numbers

### LINKED ISSUE #1340: Add option to flag uninitialized integers as UB
The reference currently states that this code is UB, but Miri deliberately accepts it:
```rust
fn main() {
    let _val = unsafe { std::mem::MaybeUninit::<usize>::uninit().assume_init() };
}
```
The reason we accept it is that the lang-team is not sure if we really want this to be UB (also see https://github.com/rust-lang/unsafe-code-guidelines/issues/71), so the conservative choice for the reference is to make it UB for now, but in Miri I fear this might lead to too many errors that people could consider false positives (even though by the letter of the reference they are true positives).

Still it could be interesting to see how much code (that Miri can run) actually fails when considering uninitialized integers UB, so having a flag in Miri to enable stricter checking would be interesting.

### PR DESCRIPTION
**Pull request by camelid**
Closes #1340.

Companion rustc PR that implements this in the Miri engine: rust-lang/rust#88670

r? @RalfJung

### CODE REVIEW

File: tests/run-pass/uninit_integer_no_flag.rs [hunk 1]
```diff
@@ -0,0 +1,6 @@
+// This test is from https://github.com/rust-lang/miri/issues/1340#issue-600900312.
+// This test passes because -Zmiri-check-number-initialization is not passed.
+
+fn main() {
+    let _val = unsafe { std::mem::MaybeUninit::<usize>::uninit().assume_init() };
```
Comment (RalfJung) [inline tests/run-pass/uninit_integer_no_flag.rs:5]: Please merge the run-pass tests into one. (There is a non-negligible constant cost per test.)

File: src/bin/miri.rs [hunk 2]
```diff
@@ -313,6 +313,9 @@ fn main() {
                 "-Zmiri-symbolic-alignment-check" => {
                     miri_config.check_alignment = miri::AlignmentCheck::Symbolic;
                 }
+                "-Zmiri-check-number-initialization" => {
```
Comment (RalfJung) [inline src/bin/miri.rs:316]: Could you add docs for this flag to the readme?

Also, the flag should affect raw pointers as well as integers and floats. Not sure how to best reflect that in its name though...

Comment (RalfJung) [reply src/bin/miri.rs:316]: Ah, actually looks like we already require raw ptrs to be init. I will have to check if we have a test for that.

File: tests/run-pass/uninit_integer_no_flag.rs (hunk 1 above)
Comment (RalfJung) [reply tests/run-pass/uninit_integer_no_flag.rs:5]: Also judging from `threadleak_ignored`, maybe a better name for the test would be `uninit_integer_ignored`.

Comment (camelid) [reply tests/run-pass/uninit_integer_no_flag.rs:5]: > Also judging from `threadleak_ignored`, maybe a better name for the test would be `uninit_integer_ignored` or `uninit_number_ignored`.

Although `threadleak_ignored` is a test of `-Zmiri-ignore-leaks`, not a test of no flag passed. But `ignored` might be more descriptive anyway.

File: tests/compile-fail/uninit_raw_ptr.rs [hunk 3]
```diff
@@ -0,0 +1,4 @@
+fn main() {
+    let _val = unsafe { std::mem::MaybeUninit::<*const str>::uninit().assume_init() };
```
Comment (RalfJung) [inline tests/compile-fail/uninit_raw_ptr.rs:2]:
```suggestion
    let _val = unsafe { std::mem::MaybeUninit::<*const u8>::uninit().assume_init() };
```
`str` is an unsized type, making it rather special, so let's use a more standard type here.

**Comment by camelid** [comment]
By the way, let me know if/when you want me to squash :)

**Comment by RalfJung** [comment]
Yeah this looks ready, so please squash. :)

File: README.md [hunk 4]
```diff
@@ -199,6 +199,8 @@ up the sysroot.  If you are using `miri` (the Miri driver) directly, see the
 Miri adds its own set of `-Z` flags, which are usually set via the `MIRIFLAGS`
 environment variable:
 
+* `-Zmiri-check-number-validity` enables checking of integer and float
+  validity (e.g., initialization) as part of enforcing validity invariants.
```
Comment (RalfJung) [inline README.md:203]:
```suggestion
  validity (e.g., they must be initialized and not carry pointer provenance) as part
  of enforcing validity invariants. This has no effect when
  `-Zmiri-disable-validation` is present.
```

**Comment by RalfJung** [comment]
@camelid could you rebase on top of master and then also update the `rust-version` file so that it picks up your rustc changes?

File: README.md (hunk 4 above)
Comment (camelid) [reply README.md]: Done!

**Comment by camelid** [comment]
> This has no effect when `-Zmiri-disable-validation` is present.

Should I add a test for this, or is it not necessary?

**Comment by RalfJung** [comment]
> Should I add a test for this, or is it not necessary?

I don't think that is necessary.

**Comment by camelid** [comment]
Thanks again for your help with this change! ❤️

**Comment by camelid** [comment]
I'll try to open a submodule sync later today.

**Pull request merged**